src/runtime/inspector_client.cpp

```diff
@@ -311,3 +330,384 @@ void TrInspectorClient::onReset()
+    std::string result;
+
+    for (size_t i = 0; i < input.size(); i += 3)
+    {
+      uint32_t val = 0;
+      int padding = 0;
+
+      for (int j = 0; j < 3; j++)
+      {
+        val <<= 8;
+        if (i + j < input.size())
+        {
+          val |= input[i + j];
+        }
+        else
+        {
+          padding++;
+        }
+      }
+
+      for (int j = 0; j < 4; j++)
+      {
+        if (j < 4 - padding)
+        {
+          result += chars[(val >> (18 - j * 6)) & 0x3F];
+        }
+        else
+        {
+          result += '=';
+        }
+      }
+    }
+
+    return result;
+  };
+
+  return base64Encode(hash);
+}
+
+void TrInspectorClient::handleWebSocketFrame()
+{
+  if (buffer_.size() < 2)
+  {
+    return; // Not enough data for frame header
+  }
+
+  // Basic WebSocket frame parsing (simplified for text frames)
+  uint8_t firstByte = static_cast<uint8_t>(buffer_[0]);
+  uint8_t secondByte = static_cast<uint8_t>(buffer_[1]);
+
+  bool fin = (firstByte & 0x80) != 0;
+  uint8_t opcode = firstByte & 0x0F;
+  bool masked = (secondByte & 0x80) != 0;
+  uint64_t payloadLength = secondByte & 0x7F;
+
+  size_t frameHeaderSize = 2;
+
+  // Handle extended payload length
+  if (payloadLength == 126)
+  {
+    if (buffer_.size() < 4)
+      return;
+    payloadLength = (static_cast<uint8_t>(buffer_[2]) << 8) | static_cast<uint8_t>(buffer_[3]);
+    frameHeaderSize = 4;
+  }
+  else if (payloadLength == 127)
+  {
+    if (buffer_.size() < 10)
+      return;
+    // For simplicity, we don't handle 64-bit payload lengths
+    DEBUG(LOG_TAG_ERROR, "WebSocket frame with 64-bit payload length not supported");
+    shouldClose_ = true;
+    return;
+  }
+
+  if (masked)
+  {
+    frameHeaderSize += 4; // mask key
+  }
+
+  if (buffer_.size() < frameHeaderSize + payloadLength)
+  {
+    return; // Not enough data for complete frame
+  }
+
+  // Handle close frame
+  if (opcode == 0x8)
+  {
+    shouldClose_ = true;
+    return;
+  }
+
+  // Handle text frame
+  if (opcode == 0x1)
+  {
+    std::vector<char> payload(payloadLength);
+    size_t payloadStart = frameHeaderSize;
+
+    if (masked)
+    {
+      // Extract mask key
+      uint8_t maskKey[4];
+      for (int i = 0; i < 4; i++)
+      {
+        maskKey[i] = static_cast<uint8_t>(buffer_[frameHeaderSize - 4 + i]);
+      }
+
+      // Unmask payload
+      for (size_t i = 0; i < payloadLength; i++)
+      {
+        payload[i] = buffer_[payloadStart + i] ^ maskKey[i % 4];
+      }
+    }
+    else
+    {
+      std::copy(buffer_.begin() + payloadStart, buffer_.begin() + payloadStart + payloadLength, payload.begin());
+    }
+
+    std::string message(payload.begin(), payload.end());
+    DEBUG(LOG_TAG_INSPECTOR, "Received WebSocket message: %s", message.c_str());
+
+    // For now, just echo the message back (placeholder for CDP implementation)
+    sendWebSocketMessage("Echo: " + message);
```

Comment:
This should be moved to `TrInspector` such as a lifetime `onMessage(client)`, and you can does this echo logic there :)